server/src/test/java/org/elasticsearch/repositories/blobstore/ChunkedBlobOutputStreamTests.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.repositories.blobstore;
+
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.MockBigArrays;
+import org.elasticsearch.common.util.MockPageCacheRecycler;
+import org.elasticsearch.indices.breaker.NoneCircuitBreakerService;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.zip.CRC32;
+import java.util.zip.CheckedOutputStream;
+
+public class ChunkedBlobOutputStreamTests extends ESTestCase {
+
+    private BigArrays bigArrays;
+
+    @Override
+    public void setUp() throws Exception {
+        bigArrays = new MockBigArrays(new MockPageCacheRecycler(Settings.EMPTY), new NoneCircuitBreakerService());
+        super.setUp();
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+    }
+
+    public void testSuccessfulChunkedWrite() throws IOException {
+        final long chunkSize = randomLongBetween(10, 1024);
+        final CRC32 checksumIn = new CRC32();
+        final CRC32 checksumOut = new CRC32();
+        final CheckedOutputStream out = new CheckedOutputStream(OutputStream.nullOutputStream(), checksumOut);
+        final AtomicLong writtenBytesCounter = new AtomicLong(0L);
+        final long bytesToWrite = randomLongBetween(chunkSize - 5, 1000 * chunkSize);
+        long written = 0;
+        try (ChunkedBlobOutputStream<Integer> stream = new ChunkedBlobOutputStream<>(bigArrays, chunkSize) {
+
+            private final AtomicInteger partIdSupplier = new AtomicInteger();
+
+            @Override
+            protected void flushBuffer() throws IOException {
+                final BytesReference bytes = buffer.bytes();
+                bytes.writeTo(out);
+                writtenBytesCounter.addAndGet(bytes.length());
+                finishPart(partIdSupplier.incrementAndGet());
+            }
+
+            @Override
+            protected void onCompletion() throws IOException {
+                if (buffer.size() > 0) {
```

Comment:
This should be handled by ChunkedBlobOutputStream itself I think

Reply:
Unfortunately, I couldn't find a neat way of encapsulating this (without adding even more complexity at least) because of the way the `onCompletion` will do a normal write if nothing has been buffered where it needs access to the buffer anyway.